### ISSUE TITLE
[RFC]: Ulysses-SP Constraints Solution

### ISSUE THREAD

**Issue by wtomin** [vllm-project/vllm-omni#1324]
### Motivation.

It is commonly known that Ulysses-SP shards the long sequence to multiple devices, and use alltoall to gather the full sequence before computing attention:

```text
(bs, seq_len/P, head_cnt, head_size) -> (bs, seq_len, head_cnt/P, head_size)
```
There are two major constraints of the Ulysses-SP implementation currently in vLLM-Omni:

1. the sequence lengths `seq_len` should be divisible by `sp_size`
2. the number of heads `head_cnt` should be divisible by `sp_size`

There is no special solution now in vLLM-Omni to the second constraint. There will throw an error if `head_cnt` is non-divisble by `sp_size`.

For the first constrain, the current solution is to use padding (pad the input sequence to the length divisible by `sp_size`) and the `hidden_states_mask` to exclude the padded tokens from attention.

See the [`SequenceParallelInput(auto_pad=True)`](https://github.com/vllm-project/vllm-omni/blob/963b64e7aec50529a476b725750b7f274c09e0f1/vllm_omni/diffusion/hooks/sequence_parallel.py#L318-L323) and [`hidden_states_mask` usage in qwen-image](https://github.com/vllm-project/vllm-omni/blob/963b64e7aec50529a476b725750b7f274c09e0f1/vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py#L991).

However this auto_padding approach brings other constraints:
1. The attention backends must support `attention_mask`: currently only `TORCH_SDPA`, `FLASH_ATTN`
2. Ring attention does not support `attention_mask`, therefore in the case of auto_padding, ulysses + ring is not feasible.
3. Many diffusers transformer models do not support `hidden_states_mask`. To support `auto_padding`, instrusive editing to support `hidden_states_mask` is required. Relevant to #1250 and #1279.

### Proposed Change.

As per discussed with @ZJY0516 , we find that cache-dit's [UAA: Ulysses Anything Attention](https://cache-dit.readthedocs.io/en/latest/user_guide/CONTEXT_PARALLEL/#uaa-ulysses-anything-attention) might be a better solution than auto_padding:

1. It requires zero_padding on the sequence. They added an extra all-gather op for scalar types to gather the seq_len value of each rank (each rank can have different sub-seq len). They claimed this extra all_gather op introduced minimal overhead.
2. It also supports arbitrary numbers of heads via additional padding and unpadding operations implemented before and after all-to-all. They claimed the overhead incurred by these extra padding and unpadding steps can be partially hidden through asynchronous communication. 

[Design Doc](https://docs.google.com/document/d/12YxSsVeD1jvL-InClkeAEnZyWFDndz_65JmXvsamuV4/edit?tab=t.0#heading=h.4ef4szrsgspp) is not ready yet. This RFC is currently for discussion only.

### CC List.

@ZJY0516 @hsliuustc0106 @SamitHuang  @mxuax @dongbo910220 @gcanlin 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by ZJY0516** [comment]
Not supporting certain resolutions is fine by me. People usually only work with a few common resolutions anyway. As long as we provide clear, reasonable error messages for unsupported cases, that should be sufficient.

**Comment by dongbo910220** [comment]
> Not supporting certain resolutions is fine by me. People usually only work with a few common resolutions anyway. As long as we provide clear, reasonable error messages for unsupported cases, that should be sufficient.

Just to clarify gently: the issue isn’t really about uncommon resolutions. Ulysses requires the total seq_len to be divisible by sp_size, and for Z‑Image seq_len = image_seq_len + text_len (with image_seq_len = (H/16)*(W/16) and text_len depending on the prompt). So even at common resolutions, prompt length variation can still trigger errors. Also, Z‑Image has n_heads = 30, so sp_size = 4/8 will always fail regardless of resolution.

**Comment by wtomin** [comment]
Yes, currently vLLM-Omni only shards the image latent, and repeats the text embedding in each rank. If Ulysses SP has no constraint on the sequence length, we may shard the text embedding too.

**Comment by dongbo910220** [comment]
**Design Doc** for this feature is now available for review:
[UAA Feature Design Doc](https://docs.google.com/document/d/1xgIPiBSxntWyUgLoDeTbPIUXNmNNF1J5EMtrxpNgUpE/edit?tab=t.0#heading=h.yk5awf1biano)

Feedback and comments are welcome!

**Comment by hsliuustc0106** [comment]
> **Design Doc** for this feature is now available for review: [UAA Feature Design Doc](https://docs.google.com/document/d/1xgIPiBSxntWyUgLoDeTbPIUXNmNNF1J5EMtrxpNgUpE/edit?tab=t.0#heading=h.yk5awf1biano)
> 
> Feedback and comments are welcome!

make sense to me, after this pr, we may solve many current issues

**Comment by wtomin** [comment]
I have read the design doc and had a few questions&suggestions:
1. What is the default value for `ulysses_mode`? Is it `strict`? ( which means that both seq_len and num_heads must be divisible by sp_size)
```python
parallel_config = DiffusionParallelConfig(
    ulysses_degree=4,
    ring_degree=2,                 # Supports Hybrid
    ulysses_mode="advanced_uaa",   # Explicit opt-in required
)
```
2. What are the valid inputs for `ulysses_mode`? besides `strict`, `advanced_uaa`, is there also `auto_pad`, or `rounding` (means to round the generated image/video resolution to meet the requirements of divisibility)
3. How to deal with the existing models with `auto_pad` supports? Do you plan to adapt them to the UAA?
4. Although the extra overhead of UAA should be small compared with ulysses-sp, we still need a test script in the CI to monitor its communication cost and performance degradation

@dongbo910220

**Comment by wtomin** [comment]
@dongbo910220  Hi, do you have any updates about your design doc? Just wondering if you have plan to raise a PR for UAA.

**Comment by dongbo910220** [comment]
> [@dongbo910220](https://github.com/dongbo910220) Hi, do you have any updates about your design doc? Just wondering if you have plan to raise a PR for UAA.

Thanks for the detailed design doc feed back. Update design doc  and the [PR](https://github.com/vllm-project/vllm-omni/pull/1379) for UAA. 

Here are the clarifications on questions:
1. Default ulysses_mode: the default is "strict". In strict mode we enforce the usual divisibility constraints (sequence must be evenly shardable for SP, and Ulysses requires num_heads divisible by ulysses_degree).
2. Valid ulysses_mode values: currently only "strict" and "advanced_uaa". auto_pad is not a ulysses_mode (it’s a per-tensor option in the model _sp_plan), and “rounding” image/video resolution is a model/pipeline choice rather than an SP mode.
3. Existing models using auto_pad: since advanced_uaa covers the same non-divisible sequence/head cases without relying on attention_mask, we plan to migrate existing auto_pad users to advanced_uaa and then deprecate auto_pad.
4.  CI perf monitoring: we added a CI perf-smoke to track UAA comm overhead (tests/diffusion/distributed/test_ulysses_uaa_perf.py). On a 4×V100 run (max across ranks, ms):
 - Pure Ulysses (ulysses=4, ring=1): strict 2.320, advanced_uaa 2.712 (ratio 1.169x, +0.392)
 - Hybrid (ulysses=2, ring=2): strict 1.612, advanced_uaa 2.213 (ratio 1.372x, +0.600)

**Comment by wtomin** [comment]
Close this issue with #1379.
